### PR TITLE
Initial commits for module django-esidoc

### PR DESCRIPTION
**Pull request by ThomSawyer**

### CODE REVIEW

File: django_esidoc/middleware.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+            uai_number = validate_ticket(request, cas_ticket)
+
+            if uai_number == request.session.get('uai_number'):
+                user = authenticate(uai_number=uai_number)
+                if user:
+                    login(request, user)
+
+            url = request.GET.get('next_url', ESIDOC_DEFAULT_REDIRECT)
+            return HttpResponseRedirect(url)
```
[the remark this answers is not in the document]
Comment (ThomSawyer) [reply django_esidoc/middleware.py]: Fixed by b38b6c5
Agreed. This is way much better.

File: django_esidoc/utils.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+def get_cas_login_url(request):
+    """Returns the CAS login url"""
+
+    client = _get_cas_client(request)
+
+    return client.get_login_url()
+
+
+def get_cas_logout_url(request):
```
[the remark this answers is not in the document]
Comment (ThomSawyer) [reply django_esidoc/utils.py]: Fixed by b38b6c5

File: django_esidoc/utils.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    client = _get_cas_client(request)
+    base_url = client.get_logout_url()
+    redirect_url = _get_redirect_url(request)
+    url = '{}?service={}'.format(base_url, redirect_url)
+
+    return url
+
+
+def validate_ticket(request, cas_ticket):
```
[the remark this answers is not in the document]
Comment (ThomSawyer) [reply django_esidoc/utils.py]: Fixed by b38b6c5

File: django_esidoc/utils.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    client = CASClient(
+        version=2,
+        server_url=server_url,
+        service_url=service_url
+    )
+
+    return client
+
+
+def get_cas_login_url(request):
```
[the remark this answers is not in the document]
Comment (ThomSawyer) [reply django_esidoc/utils.py]: Fixed by b38b6c5

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+from setuptools import setup
+
+setup(name='django-esidoc',
+      version='0.0.1.dev0',
+      description='Handle login and ticket validation for e-sidoc',
+      url='https://github.com/ThomSawyer/django-esidoc',
+      author='Thomas Carbillet',
+      author_email='thomas.carbillet@gmail.com',
+      license='GNU',
```
[the remark this answers is not in the document]
Comment (ThomSawyer) [reply setup.py]: Fixed by 8e7b84f

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
+from setuptools import setup
+
+setup(name='django-esidoc',
+      version='0.0.1.dev0',
+      description='Handle login and ticket validation for e-sidoc',
+      url='https://github.com/ThomSawyer/django-esidoc',
```
[the remark this answers is not in the document]
Comment (ThomSawyer) [reply setup.py]: Fixed by 8e7b84f

File: tests/test_middleware.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+ESIDOC_DEFAULT_REDIRECT = getattr(settings, 'ESIDOC_DEFAULT_REDIRECT', '/')
+
+
+class TestCASMiddleware(object):
+
+    uai_number = '000000A'
+
+    @staticmethod
+    def _mock_request(query_params='?'):
```
[the remark this answers is not in the document]
Comment (ThomSawyer) [reply tests/test_middleware.py]: Fixed by d367352

File: tests/test_backends.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        # THEN
+        assert user is not None
+
+    def test_get_user_fails(self):
+        """
+        Test the case where get_user couldn't find the corresponding user in the database
+        """
+        # WHEN
+        backend = CASBackend()
+        user = backend.get_user(user_id=8)
```
[the remark this answers is not in the document]
Comment (ThomSawyer) [reply tests/test_backends.py]: Fixed by d367352

File: tests/test_backends.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+ESIDOC_DEFAULT_REDIRECT = getattr(settings, 'ESIDOC_DEFAULT_REDIRECT', '/')
+
+
+class TestCASBackend(object):
+
+    uai_number = '000000A'
+
+    @pytest.fixture
+    def mock_user(self, django_user_model):
```
[the remark this answers is not in the document]
Comment (ThomSawyer) [reply tests/test_backends.py]: Fixed by d367352

File: django_esidoc/middleware.py [hunk 10]
```diff
@@ -4,9 +4,11 @@
 from django.conf import settings
 from django.contrib.auth import login, authenticate
 
+from xml.etree import ElementTree
+from xml.etree.ElementTree import ParseError
+
 from .utils import (
```
Comment (cirotix) [inline django_esidoc/middleware.py]: Can be on one line

File: django_esidoc/utils.py [hunk 11]
```diff
@@ -18,15 +19,15 @@ def _get_cas_base_url(uai_number):
...
     """Create a CAS client"""
 
     uai_number = request.session.get('uai_number')
     server_url = _get_cas_base_url(uai_number)
 
-    redirect_url = _get_redirect_url(request)
     next_page = request.path
-    service_url = '{}?next_url={}'.format(redirect_url, next_page)
+    redirect_url = get_redirect_url(request, next_page)
+    service_url = '{}'.format(redirect_url)
```
Comment (cirotix) [inline django_esidoc/utils.py]: This seems unnecessary ;)

File: tests/conftest.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+from django.contrib.sessions.middleware import SessionMiddleware
+from django.test import RequestFactory
+
+from factories import (UserFactory, InstitutionFactory)
+
+
+@pytest.fixture
+def user(uai_number):
+    user = UserFactory(
+        id=1,
```
Comment (cirotix) [inline tests/conftest.py]: I don't think that it is a good practice to fix the id.

File: tests/conftest.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        email='noel@flantier.com',
+        is_active=True
+    )
+    InstitutionFactory(
+        uai=uai_number,
+        institution_name='Lycée John, Doe',
+        user_id=1
```
Comment (cirotix) [inline tests/conftest.py]: If we are not fixing the id for the user, couldn't we use
```python
user=user
```
instead

File: tests/conftest.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+from factories import (UserFactory, InstitutionFactory)
+
+
+@pytest.fixture
+def user(uai_number):
+    user = UserFactory(
+        id=1,
+        email='noel@flantier.com',
+        is_active=True
```
Comment (cirotix) [inline tests/conftest.py]: This should go in the factory

File: tests/conftest.py [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+from django.test import RequestFactory
+
+from factories import (UserFactory, InstitutionFactory)
+
+
+@pytest.fixture
+def user(uai_number):
+    user = UserFactory(
+        id=1,
+        email='noel@flantier.com',
```
Comment (cirotix) [inline tests/conftest.py]: This should go in the factory, with a sequence to increment the email (in order to have some unique emails).

Like this
```python
email = factory.Sequence(lambda n: 'person{0}@example.com'.format(n))
```

**Comment by ThomSawyer** [comment]
Write missing tests: https://codecov.io/gh/briefmnews/django-esidoc/src/master/django_esidoc/middleware.py#L45

File: django_esidoc/middleware.py (hunk 10 above)
Comment (ThomSawyer) [reply django_esidoc/middleware.py]: Fixed by 46dab54

File: django_esidoc/utils.py (hunk 11 above)
Comment (ThomSawyer) [reply django_esidoc/utils.py]: Fixed by c5d5fdf

File: tests/conftest.py (hunk 12 above)
Comment (ThomSawyer) [reply tests/conftest.py]: Fixed by a37b1fd

File: tests/conftest.py (hunk 13 above)
Comment (ThomSawyer) [reply tests/conftest.py]: Fixed by a37b1fd

File: tests/conftest.py (hunk 14 above)
Comment (ThomSawyer) [reply tests/conftest.py]: Fixed by a37b1fd

File: tests/conftest.py (hunk 15 above)
Comment (ThomSawyer) [reply tests/conftest.py]: Fixed by a37b1fd

File: django_esidoc/middleware.py (hunk 10 above)
Comment (cirotix) [reply django_esidoc/middleware.py]: Uhm, does not seem so.

Comment (cirotix) [reply django_esidoc/middleware.py]: But it doesn't really matter as it will be blacked!

Comment (ThomSawyer) [reply django_esidoc/middleware.py]: Fixed by c808d83

**Pull request merged**